File: 3-LongestSubstringWithoutRepeatingCharacters/3-LongestSubstringWithoutRepeatingCharacters.py
# Last Updated: 2/5/2026, 9:44:13 PM
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        maxi=0
        left=0
        charSet={}
        for right,char in enumerate(s):
            if char in charSet and charSet[char]>=left:
                left=charSet[char]+1
            charSet[char]=right
            maxi=max(maxi,right-left+1)
        return maxi
                

File: 3-LongestSubstringWithoutRepeatingCharacters/3-LongestSubstringWithoutRepeatingCharacters.cpp
// Last Updated: 2/5/2026, 9:44:16 PM
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
    int hash[256];
    fill(hash, hash + 256, -1);  // Initialize all characters as unseen

    int maxLen = 0;
    int l = 0;  // Left boundary of window

    for (int r = 0; r < s.size(); ++r) {
        if (hash[s[r]] >= l) {
            l = hash[s[r]] + 1;  // Move left boundary past the last occurrence
        }
        hash[s[r]] = r;  // Update last seen index
        maxLen = max(maxLen, r - l + 1);  // Update max length
    }

    return maxLen;
    }
};